extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java

```diff
@@ -0,0 +1,571 @@
+    TaskLocation location = TaskLocation.create(
+        peonPod.getStatus().getPodIP(),
+        DruidK8sConstants.PORT,
+        DruidK8sConstants.TLS_PORT,
+        node.isEnableTlsPort()
+    );
+    client.waitForProcessToStart(location, 10, TimeUnit.MINUTES);
+
+    updateLocation(task, location);
+    updateStatus(task, TaskStatus.running(task.getId()));
+
+    JobResponse response = client.waitForJobCompletion(
+        k8sTaskId,
+        KubernetesTaskRunnerConfig.toMilliseconds(k8sConfig.maxTaskDuration),
+        TimeUnit.MILLISECONDS
+    );
+    response.setLocation(location);
+    return response;
+  }
+
+  @VisibleForTesting
+  void updateStatus(Task task, TaskStatus status)
+  {
+    TaskRunnerUtils.notifyStatusChanged(listeners, task.getId(), status);
+  }
+
+  @VisibleForTesting
+  void updateLocation(Task task, TaskLocation location)
+  {
+    TaskRunnerUtils.notifyLocationChanged(listeners, task.getId(), location);
+  }
+
+  @Override
+  public void shutdown(String taskid, String reason)
+  {
+    client.cleanUpJob(new K8sTaskId(taskid));
+  }
+
+
+  @Override
+  public Optional<InputStream> streamTaskReports(String taskid)
+  {
+    return Optional.absent();
+  }
+
+  @Override
+  public List<Pair<Task, ListenableFuture<TaskStatus>>> restore()
+  {
+    return ImmutableList.of();
+  }
+
+  @Override
+  public void start()
+  {
+    log.info("Starting cleanup executor for jobs older than 1 day....");
+    cleanupExecutor.scheduleAtFixedRate(
+        () ->
+            client.cleanCompletedJobsOlderThan(
+                KubernetesTaskRunnerConfig.toMilliseconds(k8sConfig.taskCleanupDelay),
+                TimeUnit.MILLISECONDS
+            ),
+        1,
+        KubernetesTaskRunnerConfig.toMilliseconds(k8sConfig.taskCleanupInterval),
+        TimeUnit.MILLISECONDS
+    );
+    log.info("Started cleanup executor for jobs older than 1 day....");
+  }
+
+
+  @Override
+  public void stop()
+  {
+    log.info("Stopping...");
+    cleanupExecutor.shutdownNow();
+    log.info("Stopped...");
+  }
+
+  @Override
+  public Map<String, Long> getTotalTaskSlotCount()
+  {
+    return ImmutableMap.of("taskQueue", (long) taskQueueConfig.getMaxSize());
+  }
+
+  private List<String> javaOpts(Task task)
+  {
+    final List<String> javaOpts = new ArrayList<>();
+    Iterables.addAll(javaOpts, k8sConfig.javaOptsArray);
+
+    // Override task specific javaOpts
+    Object taskJavaOpts = task.getContextValue(
+        ForkingTaskRunnerConfig.JAVA_OPTS_PROPERTY
+    );
+    if (taskJavaOpts != null) {
+      Iterables.addAll(
+          javaOpts,
+          new QuotableWhiteSpaceSplitter((String) taskJavaOpts)
+      );
+    }
+
+    javaOpts.add(StringUtils.format("-Ddruid.port=%d", DruidK8sConstants.PORT));
+    javaOpts.add(StringUtils.format("-Ddruid.plaintextPort=%d", DruidK8sConstants.PORT));
+    javaOpts.add(StringUtils.format("-Ddruid.tlsPort=%d", node.isEnableTlsPort() ? DruidK8sConstants.TLS_PORT : -1));
```

Comment:
Should these ports be hardcoded? Which ports are these? The listen ports on the pod which the OL uses? Or, the listen ports on Druid which the K8s peon uses? If for OL, should we use the configured ports? If for the pod, can we guarantee these ports are always available?

Reply:
this is for the peon task.  For the MM based peon tasks, these were dynamic because you were running multiple peons on a host. So the host was static, and you had to make the ports dynamic. 

Now in the k8s world, each job pod will have a unique host, so we can use a static port (we don't need it dynamic to avoid collisions).